### PR TITLE
Admin panel

### PR DESCRIPTION
**Pull request by DavidBabiy**
- Added endpoint for getting all courses, and and disabling it by id
- Added endpoint for getting all groups, and and disabling it by id
- Added endpoints for getting all users, and disabling user by id
- Created frontend admin panel
- Created advice controller and one custom exception

### CODE REVIEW

**Review by nazarchykur — APPROVED** [review]

File: src/main/java/com/softserve/itacademy/dto/CourseDto.java [hunk 1]
```diff
@@ -1,4 +1,4 @@
-package com.softserve.itacademy.entity.dto;
+package com.softserve.itacademy.dto;
 
 import com.softserve.itacademy.entity.Course;
```
Comment (SerhiiTsoi) [inline src/main/java/com/softserve/itacademy/dto/CourseDto.java:3]: все таки подумайте на счет Builder, мне кажется так будет красивее))

Comment (SerhiiTsoi) [reply src/main/java/com/softserve/itacademy/dto/CourseDto.java:3]: так жу думаю стоит написать equals hashcode include pkey

**Review by SerhiiTsoi — APPROVED** [review]

**Review by juliamartyn — APPROVED** [review]

**Review by MykhailoIvanyshyn — APPROVED** [review]

File: src/main/java/com/softserve/itacademy/controller/CourseController.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+import static org.springframework.http.HttpStatus.OK;
+
+@RestController
+@RequestMapping("/api/courses")
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/controller/CourseController.java:14]: You didn't read carefully the resources I sent in chat. How about versioning of the API?

File: src/main/java/com/softserve/itacademy/entity/Group.java [hunk 3]
```diff
@@ -15,9 +15,10 @@
 @Entity
 @Table(name = "student_groups")
 public class Group extends BasicEntity {
-    private String name;
 
+    private String name;
     private Integer ownerId;
+    private Boolean disabled = false;
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/entity/Group.java:21]: How about using primitive here. It has already default false value

File: src/main/java/com/softserve/itacademy/exception/handler/GenericExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+import static org.springframework.http.HttpStatus.NOT_FOUND;
+
+@Slf4j
+@ControllerAdvice
+public class GenericExceptionHandler {
+
+    @ExceptionHandler(NotFoundException.class)
+    public ResponseEntity<Object> handleAccessDeniedException(NotFoundException exception) {
+        log.debug(exception.getMessage());
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/exception/handler/GenericExceptionHandler.java:17]: wrong log level

File: src/main/java/com/softserve/itacademy/exception/handler/GenericExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        return new ResponseEntity<>(exception.getMessage(), NOT_FOUND);
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/exception/handler/GenericExceptionHandler.java:18]: It's a bad practice to respond with an exception message

File: src/main/java/com/softserve/itacademy/service/implamentations/CourseServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    @Override
+    public void updateDisabled(Integer id, Boolean disabled) {
+        Course group = getById(id);
+        group.setDisabled(disabled);
+        courseRepository.save(group);
+    }
+
+    private Course getById(Integer id) {
+        return courseRepository.findById(id).orElseThrow(NotFoundException::new);
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/service/implamentations/CourseServiceImpl.java:40]: I don't like the approach of handling not found entities. Let's discuss this during the next meeting

File: src/main/java/com/softserve/itacademy/service/implamentations/GroupServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    @Override
+    public void delete(Integer id) {
+        groupRepository.delete(getById(id));
+    }
+
+    @Override
+    public void updateDisabled(Integer id, Boolean disabled) {
+        Group group = getById(id);
+        group.setDisabled(disabled);
+        groupRepository.save(group);
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/service/implamentations/GroupServiceImpl.java:36]: Wrong approach. You don't have to select and then save the whole object updating only one attribute. You are wasting the resources Instead just add appropriate method to the repo with the custom query (Update table set disabled=:some_param) and use this method whenever you need

File: src/main/java/com/softserve/itacademy/service/implamentations/UserServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    public List<UserDto> findAll() {
+        return userRepository.findAll().stream().map(UserDto::create).collect(Collectors.toList());
+    }
+
+
+    @Override
+    public void updateDisabled(Integer id, Boolean disabled) {
+        User user = userRepository.findById(id).orElseThrow(NotFoundException::new);
+        user.setDisabled(disabled);
+        userRepository.save(user);
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/service/implamentations/UserServiceImpl.java:33]: the same here

File: src/main/java/com/softserve/itacademy/controller/GroupController.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        return new ResponseEntity<>(groupService.findAll(), HttpStatus.OK);
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Void> delete(@PathVariable Integer id) {
+        groupService.delete(id);
+        return new ResponseEntity<>(OK);
+    }
+
+    @DeleteMapping("/{id}/disabled")
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/controller/GroupController.java:34]: This is not correctly designed endpoint.
Use PATCH instead to update one filed. Why did you use DELETE?

File: src/main/java/com/softserve/itacademy/controller/GroupController.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    public ResponseEntity<Void> updateDisabled(@PathVariable Integer id, @RequestParam Boolean disabled) {
```
Comment (bkomar) [inline src/main/java/com/softserve/itacademy/controller/GroupController.java:35]: It is better to use json body for updating disabled attribute as it is a part of the resource
Read this https://stackoverflow.com/questions/28459418/use-of-put-vs-patch-methods-in-rest-api-real-life-scenarios

**Pull request merged**